### DISCUSSION TITLE
Planned Features

### DISCUSSION THREAD

**Discussion by vincode-io** [vincode-io/Zavala#56] [Ideas]
This is to discuss the [Planned Features](https://github.com/vincode-io/Zavala/wiki/Planned-Features) for Zavala.  If you see anything missing or that you disagree with, here is the place to discuss it.

**Comment by kerim** [comment]
Glad to see this project. Strongly urge you to consider having backlinks and transclusion included in the app from the beginning, rather than as something to be added later on. I think the app will be much stronger if the implementation of these features is foundational - part of the overall design and philosophy of the app - rather than an afterthought. One reason I use Roam now, rather than Dynalist (which is what I used before) is because of how well and deeply these features are integrated into Roam. I find many implementations of these features in other apps quite disappointing, although Workflowy has now done a good job with transclusion (what they call "Mirrors").

Another thing I think should be there from the start, is good import and export to/from other outlining programs and text editors: OPML, Markdown, formatted text, and maybe even JSON and Roam's EDN format as well...

Search should be switchable between in-document and app-wide, and should allow boolean filtering of results.

**Reply by macosxguru** [reply]
Is this an outliner or a competitor to Roam? If you are using Roam, isn't that at its core an outliner? Why would you need an external outliner to augment Roam?

If this is the mission statement: "I would kill for a good simple outliner that synced across iOS and Mac, and had a great simple UI on all three platforms." Where does backlinks and transclusion figure in that descriptor? 

Export to OPML and Markdown makes sense but formatted text, JSON and EDN? I am not sure that is needed in a "simple" outliner. 

macosxguru

**Reply by kerim** [reply]
My point was that many outliners in 2021 now consider such features to be essential (or at least their users do) and so are going back and retrofitting their apps to include them - but they have generally done it in a ill-thought out way, so if there is the slightest chance that they are planned for later on, it is best to incorporate them from the ground up. 

I think the concept of "simple" is often misunderstood. The iPhone is simple to use because it hides its complexity, not because the complexity isn't there. Workflowy did a lot of work to make transclusion "simple" with their "mirrors" concept and I think they did a great job. Simple certainly doesn't mean simple to program. A theming system might be a huge burden to program even though it is a common enough feature these days. (I'm personally happy to do without one as long as there is a dark theme and control over the font size, but to each his own...) I trust the developer will look at all the suggestions and go with whatever best fits their own vision for the project, and what is easy for them to make and maintain given the tools they are using.

**Reply by vincode-io** [reply]
Thanks for the input.  This was the kind of discussion I was hoping to get.

Regarding transclusion, I am strongly considering it.  As you pointed out, if you ever plan to support it, you should do so from the beginning.  As a developer, transclusion looks like a pretty difficult feature to implement.  I'm not sure if this is just a power user feature and wouldn't be used by most users or if it is a must have.  I have to balance my limited development time with what features are going to be most beneficial.

**Reply by vincode-io** [reply]
@kerim Could you share with me how you use transclusion?  I'm still trying to gauge if it fits the project or not and think it would be helpful if I knew how you were using it.

**Reply by kerim** [reply]
Transclusion is one of those features that, once you have it, becomes something you use every day. Here are some examples:

1. I am a teacher and I meet regularly with students about their dissertations. I have notes on each student's topic and progress, etc. Rather than just linking to it and having to move back and forth between two notes, I include that note within my notes for each meeting with that student and update it right there. Later, when I go to their thesis note, that is there, along with backlinks to each of the meetings I had where we discussed it. 

2. Another example is in doing research, where I might have a concept I am refining, or a subject that I am collecting data. Whenever it comes up in my notes I can include the full notes for that concept or subject and update it right there. Then, again, when going back to that subject page I also have the backlinks to each place where I discussed the concept. 

As you see, for me the two tools: transclusion and backlinks, are very closely linked, each building off the other. I have discussed this in more depth in a blog post I wrote when I first started using Roam last year:

https://anthrodendum.org/2020/05/05/roam-if-you-want-to/

**Comment by macosxguru** [comment]
I would like the ability to use my own fonts. In a user-selectable size.
Themes would be a good addition. 
In full screen mode I don't want a line in the editor to span the whole screen. User-defined line width for all editing windows would be a good addition.

macosxguru

**Reply by vincode-io** [reply]
That's good feedback.  Do you think that Themes are important while you are editing the outline or most important when generating output, like print or a PDF?

**Reply by macosxguru** [reply]
I don't know the right answer to that question for the average user. For me, personally, I am interested in themes for the editing window. The way the output looks is not important to me. It is going to be eventually a part of a Markdown formatted document which is going to be handled by pandoc for the final output.

macosxguru

**Reply by mry53** [reply]
Themes during editing are more important than when generating output.
If I am going to live in/use a tool I need to be comfortable in it, and a good part of that is 'making it mine'. Same idea for controlling line length, font style and size, spacing. My eyes and my sensibilities are unique (as are everyone's), and I have moved on from many tools because they do not allow some tuning to my workflow and needs.

**Reply by vincode-io** [reply]
Thanks for the feedback. I'm not currently sure where we'll land regarding themes, but I am taking your input seriously.

**Reply by vincode-io** [reply]
Could you guys share some examples of where you think Themes have been done well? Screenshots would be helpful if you can get them.

Are we talking about a common system wide theme that would be applied to all Outlines or are we talking about a Theme catalog that can be applied to each Outline separately?

**Reply by mry53** [reply]
NVUltra (in Beta); Bear; Ulysses. 
All have theme files which apply system wide. Individual themes for each outline is probably overkill.
The NVUltra themes are from Multimarkdown Composer (same developer) and offer the capability of setting up pretty much any parameter (example attached; open/save as a text file.) 
Bear themes are pre-built. 
Ulysses themes are Xml-based; there is an active theming community for sharing.
My top choice for themes? NVUltra. Very flexible, easy to edit.

**Reply by vincode-io** [reply]
I was thinking of something a lot simpler and less powerful. I don't want users to have to edit a special file format if possible.

What about an area in the Preferences where someone can set the font and font size for each individual level? I realize that this is real limiting, but I'm trying to not hit that "paradox of choice" thing that causes people to lock up and not make any decisions.

Is that too limiting and just not worth implementing?

**Reply by mry53** [reply]
I actually agree with you about the tyranny of choice.
I was providing examples of theming I use and like, not specifically saying go that far. Certainly dark/light themes are a basic necessity these days. Beyond that, some typographic control for accessibility at minimum; and for readability and screen/eye ease (if I am staring at a screen for any length of time more than a few minutes my eyes need to be comfortable.) Per-level font size enhances "outline-ability" (i.e., ease of detecting levels.) 
Notes on individual outline items work well with different font choices as well. It is a slippery slope, though.

**Comment by mry53** [comment]
[Solarized III.txt](https://github.com/vincode-io/Zavala/files/6151684/Solarized.III.txt)

**Comment by mknudsen** [comment]
I like the overall keyboard experience on iPad. A shortcut to ad a note would be neat.

**Reply by vincode-io** [reply]
The shortcut to add a note is `^-`. To delete a note, just add the shift key modifier.

I realize this isn't very discoverable on the iPad. I'll see what I can do to make this more obvious.

**Reply by mknudsen** [reply]
Hmm, still not able to use that. Maybe that‘s due to my keyboard layout being German? I also don‘t see the command when long pressing CMD (I‘m on the iOS 15 Beta).

**Reply by vincode-io** [reply]
It won't show up under the command menu because it is using only the control modifier key.

**Comment by latvia234** [comment]
@vincode-io What do you mean with 'Tag based organization'? Does this only apply to documents/files level?


I would love tags on line level. I use tags all the time in Dynalist. 

One of the many use cases of tags on line level: https://youtu.be/wd8aRoTULoU?t=104

Are there plans to make tags available on line level?

**Reply by vincode-io** [reply]
Right now you can only organize tags at the outline level. I can see how they would be useful at the line level though and will consider them for a future feature, but don't have any implementation plans for them right now.

**Reply by latvia234** [reply]
https://github.com/vincode-io/Zavala/discussions/114

**Comment by latvia234** [comment]
In the [planned features](https://github.com/vincode-io/Zavala/wiki/Planned-Features) I can see the following:

> - Link Support
>    - Should be able to link to any Outline

There are other outliner apps that can create links to open any outline that's inside the app. However when exporting this outline to another outliner app, I had to manually go through every link to replace it with something else. Because I do not want to break my workflow. So you can say that I was locked into the app.

I am happy to see that more and more applications are using `[[wikilinks]]`. 
- Importing and exporting will be easier
- You are not locked into using one app
- Existing apps are using wikilinks so more people are familiar with it
   - Apps like [Noteplan](https://noteplan.co/), [Obsidian](https://obsidian.md/), [Roam](https://roamresearch.com/), [Athens Research](https://github.com/athensresearch/athens), [Logseq](https://logseq.com/)

Could you please make `[[wikilinks]]` available to create new outlines, or link to any existing outline?

It would be helpful if the new outline could be made instantly when you're typing the title of the wikilink. In combination with backlinks this could be really helpful when there are multiple wiklinks linking to one outline. Even if the outline itself is empty.

Could you also make it so that when you...
- update the outline title inside a wikilink, the change will be visible in all other wikilinks that are linking to the same outline?
- update the outline title it will be reflected in all wikilinks that are linking to the outline?

**Reply by vincode-io** [reply]
We don't agree with the trend to add Twitter or Wiki style syntax to outlines. We think it makes ugly documents that are hard to read.

You can create links to other outlines using the link dialog. Just activate it using the menu command or toolbar item on macOS. The link function is available on iOS using the standard Edit menu. You can then start typing in the title of the outline you are interested in. It will supply you with a menu of outlines to choose from. Select that and it will create a link in your document to the other outline.

We use markdown as our flavor of choice to preserve formatting and linking in our OPML files. I realize that isn't compatible with Wiki style outliners, so that doesn't help you much. I would like to see a standard for exporting a linked library of outlines created. If one ever comes up, we'll be one of the first applications to support it.

**Reply by latvia234** [reply]
If there's a better option instead of the Wiki style syntax, I'm happy to try it. I'm just a basic user who loves outlines and like to keeps things as simple as possible. So if markdown is a better option here I like to test it.

https://user-images.githubusercontent.com/86680397/130939685-5d3707f3-8cd2-46e3-914c-c0e2625839a7.mov

> The link function is available on iOS using the standard Edit menu.

Do you mean the first menu that you see in the video with "Copy Document Link"?

Then after testing I tried to paste the link into an outline. It creates a long link. `[description](here1)` I tried to paste it into `(here1)` but that is not possible/not easy in iOS. Or is there another option to use a Document Link inside a markdown link syntax?

**Comment by vincode-io** [comment]
This is what you are looking for:

![IMG_0064](https://user-images.githubusercontent.com/16448027/131036882-7576d2d9-ab41-4f0f-a574-ebf167a7cddf.PNG)

You might have to scroll the Edit menu to see it. I can now see that the functionality might not be very discoverable on iOS. I'll see what I can do to make that better.

**Reply by latvia234** [reply]
> This is what you are looking for:

Yes! Thanks for your help.

**Comment by funkydan2** [comment]
+1 for **Focus** on Rows. I regularly use OmniOutliner, and this is an essential feature for an outliner in my useage.
